network/src/transport.rs

```diff
@@ -207,156 +218,948 @@ pub async fn perform_handshake<T: TSocket>(
+
+        let listener_task = async move {
+            // accept one inbound connection from dialer
+
+            let (inbound, _dialer_addr) = inbounds.next().await.unwrap().unwrap();
+            let mut conn = inbound.await.unwrap();
+
+            // check connection metadata
+
+            assert_eq!(conn.metadata.peer_id, dialer_peer_id);
+            expect_ip4_tcp_noise_addr(&conn.metadata.addr);
+            assert_eq!(conn.metadata.origin, ConnectionOrigin::Inbound);
+            assert_eq!(
+                conn.metadata.messaging_protocol,
+                MessagingProtocolVersion::V1
+            );
+            assert_eq!(
+                conn.metadata.application_protocols,
+                supported_protocols_clone,
+            );
+
+            // test the socket works
+
+            let msg = write_read_msg(&mut conn.socket, b"foobar").await;
+            assert_eq!(&msg, b"barbaz".as_ref());
+
+            conn.socket.close().await.unwrap();
+        };
+
+        let dialer_task = async move {
+            // dial listener
+
+            let mut conn = dialer_transport
+                .dial(listener_addr.clone())
+                .unwrap()
+                .await
+                .unwrap();
+
+            // check connection metadata
+
+            assert_eq!(conn.metadata.peer_id, listener_peer_id);
+            assert_eq!(conn.metadata.addr, listener_addr);
+            assert_eq!(conn.metadata.origin, ConnectionOrigin::Outbound);
+            assert_eq!(
+                conn.metadata.messaging_protocol,
+                MessagingProtocolVersion::V1
+            );
+            assert_eq!(conn.metadata.application_protocols, supported_protocols);
+
+            // test the socket works
+
+            let msg = write_read_msg(&mut conn.socket, b"barbaz").await;
+            assert_eq!(&msg, b"foobar".as_ref());
+
+            conn.socket.close().await.unwrap();
+        };
+
+        rt.block_on(future::join(listener_task, dialer_task));
+    }
+
+    #[test]
+    fn test_mutual_auth_fail_dialer_unauthenticated_tcp_transport() {
+        let (
+            mut rt,
+            (_listener_peer_id, listener_key),
+            (dialer_peer_id, dialer_key),
+            trusted_peers,
+            supported_protocols,
+        ) = setup();
+
+        // remove dialer from trusted_peers set
+        trusted_peers
+            .write()
+            .unwrap()
+            .remove(&dialer_peer_id)
+            .unwrap();
+
+        let listener_transport = LibraNetTransport::new(
+            LIBRA_TCP_TRANSPORT.clone(),
+            listener_key,
+            Some(trusted_peers.clone()),
+            HANDSHAKE_VERSION,
+            supported_protocols.clone(),
+        );
+
+        let dialer_transport = LibraNetTransport::new(
+            LIBRA_TCP_TRANSPORT.clone(),
+            dialer_key,
+            Some(trusted_peers),
+            HANDSHAKE_VERSION,
+            supported_protocols,
+        );
```

Comment:
this is the same code over and over again and I suspect there's a lot more copy pasta

perhaps we could shove more of this into setup and pass in the transport or a transport new function pointer to create that type...

Reply:
managed to clean this up a lot. -300 lines : )